### PR TITLE
Use CloudFormations for AWS infrastructure

### PR DESCRIPTION
**Pull request by ironcladlou**
This commit totally rewrites the AWS infrastructure handling to use AWS
CloudFormations, replacing all the bespoke AWS API code, because it has grown
beyond our ability to maintain, and AWS can do most of this work for us.
CloudFormations also presents a more clear and declarative infrastructure
contract for end users through the CloudFormations template construct.

The main objective refactor is to achieve functional parity with the existing
behavior, but implemented in CloudFormations. This commit is totally compatible
with the existing CLI/documentation.

There are some notable structural changes introduced:

- DNS infrastruture is now better isolated. Instead of using the base domain's
zone as the public zone for the hosted cluster, a new public zone is provisioned
for the cluster and an NS record is installed into the base domain zone which
delegates to the cluster's public zone. This means the hosted cluster only knows
about zones which are designated specifically for that cluster and can no longer
leak records into the base domain zone. It also means that when cleaning up, we
only need to track the NS record in the base domain zone and can safely delete
everything in the hosted cluster's zones.

- The code itself is restructured to no longer distinguish between IAM and other
infra, with the intent of separating IAM out in a followup after a deliberate
audit of our policy needs. This means the `create iam` command is no longer
necessary and has been deleted.

- The CLI is refactored to make a clearer distinction between provisioning (or
receiving) infrastructure and installing supporting resources into that
infrastructure (e.g. OIDC resources installed into an S3 bucket, or installing
an NS record into a zone).

### CODE REVIEW

File: cmd/cluster/create.go [hunk 1]
```diff
@@ -108,14 +113,20 @@ func NewCreateCommand() *cobra.Command {
 }
 
 func CreateCluster(ctx context.Context, opts Options) error {
+	if len(opts.ReleaseImage) == 0 {
```
Comment (enxebre) [inline cmd/cluster/create.go:116]: Is this actually related to the PR?

Comment (enxebre) [reply cmd/cluster/create.go:116]: Just realised is moving code.

File: cmd/infra/aws/create.go [hunk 2]
```diff
@@ -1,225 +1,395 @@
...
 	}
-	r53client, err := route53Client(o.AWSCredentialsFile)
+	jwks, err := json.MarshalIndent(KeyResponse{Keys: keys}, "", "  ")
 	if err != nil {
-		return nil, err
+		return fmt.Errorf("failed to marshal KeyResponse: %w", err)
 	}
-	result.PublicZoneID, err = o.LookupPublicZone(r53client)
+
+	_, err = s3Client.PutObject(&s3.PutObjectInput{
```
Comment (enxebre) [inline cmd/infra/aws/create.go:297]: nit: this can be inlined.

File: cmd/infra/aws/create.go [hunk 3]
```diff
@@ -1,225 +1,395 @@
...
+		ACL:    aws.String("public-read"),
+		Body:   bytes.NewReader(jwks),
+		Bucket: aws.String(bucketName),
+		Key:    aws.String("openid/v1/jwks"),
```
Comment (enxebre) [inline cmd/infra/aws/create.go:301]: move `openid/v1/jwks` into a cosntant?

File: cmd/infra/aws/cloudformation/cluster.yaml [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+    ParameterLabels:
+      InfrastructureName:
+        default: "Infrastructure Name"
+
+Mappings:
+  RegionMap:
+    us-east-1:
+      DHCPDomainName: "ec2.internal"
+    us-west-1:
+      DHCPDomainName: "us-west-1.compute.internal"
```
Comment (csrwng) [inline cmd/infra/aws/cloudformation/cluster.yaml:36]: Does this mean that we only support `us-east-1` and `us-west-1` as regions right now?

File: cmd/infra/aws/cloudformation/cluster.yaml [hunk 5]
```diff
@@ -0,0 +1,488 @@
...
+      Tags:
+      - Key: !Join [ "", [ "kubernetes.io/cluster/", !Ref InfrastructureName ] ]
+        Value: "owned"
+
+  OIDCProvider:
+    Type: AWS::IAM::OIDCProvider
+    Properties:
+      Tags:
+      - Key: !Join [ "", [ "kubernetes.io/cluster/", !Ref InfrastructureName ] ]
+        Value: "owned"
```
Comment (csrwng) [inline cmd/infra/aws/cloudformation/cluster.yaml:52]: I thought we were going to preserve Name tags... it's hard to understand what these are in the console without a name.

File: cmd/infra/aws/cloudformation/cluster.yaml [hunk 6]
```diff
@@ -0,0 +1,488 @@
...
+
+  VPC:
+    Type: "AWS::EC2::VPC"
+    Properties:
+      EnableDnsSupport: "true"
+      EnableDnsHostnames: "true"
+      CidrBlock: "10.0.0.0/16"
+      Tags:
+      - Key: !Join [ "", [ "kubernetes.io/cluster/", !Ref InfrastructureName ] ]
+        Value: "owned"
```
Comment (csrwng) [inline cmd/infra/aws/cloudformation/cluster.yaml:256]: Name?

File: cmd/infra/aws/create.go [hunk 7]
```diff
@@ -1,225 +1,395 @@
...
+		subdomain = getStackOutput(stack, "Subdomain")
+		zoneID = getStackOutput(stack, "BaseDomainHostedZoneId")
+		if len(nameservers) > 0 {
+			return true, nil
+		}
+		return false, nil
+	}, ctx.Done())
 	if err != nil {
-		return nil, fmt.Errorf("failed to create client session: %w", err)
+		return fmt.Errorf("failed waiting for subdomain nameserver output")
```
Comment (enxebre) [inline cmd/infra/aws/create.go:363]: It'd be nice to store the real error preventing this from proceeding, e.g last error returned by `stack, err := getStack(cf, stackID)` or `nameservers is empty` and return it here

File: cmd/infra/aws/create.go [hunk 8]
```diff
@@ -1,225 +1,395 @@
...
-	}
-	awsConfig.Credentials = credentials.NewSharedCredentials(creds, "default")
-	s, err := session.NewSession(awsConfig)
+// TODO: Replace direct Route53 manipulation with CloudFormations
+func (o *CreateInfraOptions) configureSubdomain(ctx context.Context, cf *cloudformation.CloudFormation, r53 route53iface.Route53API, stackID string) error {
+	var nameservers, zoneID, subdomain string
+	err := wait.PollUntil(5*time.Second, func() (bool, error) {
+		stack, err := getStack(cf, stackID)
+		if err != nil {
+			log.Error(err, "failed to get stack", "id", stackID)
```
Comment (enxebre) [inline cmd/infra/aws/create.go:351]: can we print err in this log

File: cmd/infra/aws/create.go [hunk 9]
```diff
@@ -1,225 +1,395 @@
...
+					},
+				},
+			},
+		},
+		HostedZoneId: aws.String(zoneID),
 	}
-	awsConfig.Credentials = credentials.NewSharedCredentials(creds, "default")
-	s, err := session.NewSession(awsConfig)
+
+	_, err = r53.ChangeResourceRecordSets(params)
```
Comment (enxebre) [inline cmd/infra/aws/create.go:388]: nit: could inline this.

File: cmd/infra/aws/destroy.go [hunk 10]
```diff
@@ -33,430 +38,294 @@ func NewDestroyCommand() *cobra.Command {
...
-	errs = append(errs, o.DestroyEIPs(ctx, ec2client)...)
-	errs = append(errs, o.DestroyDNS(ctx, route53client)...)
-	return utilerrors.NewAggregate(errs)
+	log.Info("Deleted the stack", "id", *stack.StackId)
+
+	return nil
 }
 
-func (o *DestroyInfraOptions) DestroyELBs(ctx context.Context, client elbiface.ELBAPI, vpcID *string) []error {
+func deleteUnmanagedELBs(ctx context.Context, client elbiface.ELBAPI, vpcID string) error {
```
Comment (csrwng) [inline cmd/infra/aws/destroy.go:151]: Should have a TODO to delete NLBs as well. Also, I wonder if volumes need to be explicitly cleaned up as well.

File: cmd/infra/aws/destroy.go [hunk 11]
```diff
@@ -33,430 +38,294 @@ func NewDestroyCommand() *cobra.Command {
...
+	}
+	log.Info("Found stack", "id", *stack.StackId)
 
-func (o *DestroyInfraOptions) DestroyInfra(ctx context.Context) error {
-	var errs []error
-	ec2client, err := ec2Client(o.AWSCredentialsFile, o.Region)
+	// Clean up the OIDC S3 bucket so it can be deleted along with the stack
+	bucket := getStackOutput(stack, "OIDCBucketName")
+	err = emptyBucket(ctx, bucket, s3)
+	if err != nil {
```
Comment (enxebre) [inline cmd/infra/aws/destroy.go:97]: nit: could inline this.

File: cmd/infra/aws/destroy.go [hunk 12]
```diff
@@ -33,430 +38,294 @@ func NewDestroyCommand() *cobra.Command {
...
+		return fmt.Errorf("failed to empty the OIDC bucket: %w", err)
+	}
+	log.Info("Emptied OIDC bucket", "id", bucket)
+
+	// Delete the NS record from the base domain hosted zone
+	baseDomainZoneID := getStackOutput(stack, "BaseDomainHostedZoneId")
+	subdomain := getStackOutput(stack, "Subdomain")
+	err = deleteRecord(ctx, r53, baseDomainZoneID, "NS", subdomain)
+	if err != nil {
+		return fmt.Errorf("failed to clean up the base domain zone: %w", err)
```
Comment (enxebre) [inline cmd/infra/aws/destroy.go:107]: nit: could inline this.

File: cmd/infra/aws/destroy.go [hunk 13]
```diff
@@ -33,430 +38,294 @@ func NewDestroyCommand() *cobra.Command {
...
+	}
+	log.Info("Cleaned up the base domain hosted zone", "id", baseDomainZoneID, "subdomain", subdomain)
+
+	// Find and delete any non-default unmanaged DNS records
+	subdomainPrivateZoneID := getStackOutput(stack, "SubdomainPrivateZoneId")
+	err = deleteNonDefaultRecords(ctx, r53, subdomainPrivateZoneID)
+	if err != nil {
```
Comment (enxebre) [inline cmd/infra/aws/destroy.go:114]: nit: could inline this.

File: cmd/infra/aws/destroy.go [hunk 14]
```diff
@@ -33,430 +38,294 @@ func NewDestroyCommand() *cobra.Command {
...
+		return fmt.Errorf("failed to clean up the subdomain private zone: %w", err)
+	}
+	log.Info("Cleaned up the subdomain private zone", "zone", subdomainPrivateZoneID)
+	subdomainPublicZoneID := getStackOutput(stack, "SubdomainPublicZoneId")
+	err = deleteNonDefaultRecords(ctx, r53, subdomainPublicZoneID)
+	if err != nil {
```
Comment (enxebre) [inline cmd/infra/aws/destroy.go:120]: nit: could inline this.

File: cmd/infra/aws/destroy.go [hunk 15]
```diff
@@ -33,430 +38,294 @@ func NewDestroyCommand() *cobra.Command {
...
+	err = deleteUnmanagedELBs(ctx, elbclient, vpcID)
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to delete unmanaged ELBs: %w", err)
 	}
-	route53client, err := route53Client(o.AWSCredentialsFile)
+	log.Info("Cleaned up unmanaged ELBs")
+
+	// Find and delete any unmanaged leaked security groups
+	err = deleteUnmanagedSecurityGroups(ctx, ec2client, vpcID)
```
Comment (enxebre) [inline cmd/infra/aws/destroy.go:135]: nit: could inline this.

**Review by csrwng — COMMENTED** [review]
Good stuff Dan, some comments

File: cmd/infra/aws/destroy.go [hunk 16]
```diff
@@ -33,430 +38,294 @@ func NewDestroyCommand() *cobra.Command {
...
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to delete unmanaged security groups: %w", err)
 	}
-	elbclient, err := elbClient(o.AWSCredentialsFile, o.Region)
+	log.Info("Cleaned up unmanaged security groups")
+
+	// Delete the stack itself
+	err = deleteStack(ctx, cf, stack)
 	if err != nil {
```
Comment (enxebre) [inline cmd/infra/aws/destroy.go:143]: nit: could inline this.

File: cmd/infra/aws/create.go (hunk 8 above)
Comment (ironcladlou) [reply cmd/infra/aws/create.go:350]: It's already logged through the `Error` interface

File: cmd/infra/aws/cloudformation/cluster.yaml (hunk 4 above)
Comment (ironcladlou) [reply cmd/infra/aws/cloudformation/cluster.yaml:36]: Yeah, need to add appropriate mappings for any regions we want to support

File: cmd/infra/aws/cloudformation/cluster.yaml (hunk 5 above)
Comment (ironcladlou) [reply cmd/infra/aws/cloudformation/cluster.yaml:52]: Sorry, can you remind me what that tag and expected value should be?

File: cmd/infra/aws/destroy.go (hunk 10 above)
Comment (ironcladlou) [reply cmd/infra/aws/destroy.go:145]: Yeah, unfortunately we're wrestling with the following facts now:

1. k8s (and users?) can create arbitrary resources that will become dependencies of Stack resources
2. Stack resources can only be deleted by AWS

So when deleting the Stack given some unbounded/unknown future set of dependent resources associated with the Stack resources, we have a couple of options I know of:

1. Do what this PR does which is try to identify any non-stack resources and destroy them so that Stack deletion will succeed
2. Use the Stack deletion setting which deletes the Stack regardless of dependencies, orphaning the non-Stack resources

Not a great situation.

File: cmd/infra/aws/cloudformation/cluster.yaml (hunk 5 above)
Comment (csrwng) [reply cmd/infra/aws/cloudformation/cluster.yaml:52]: it's the `Name` tag and for example for the vpc, it's `[INFRA-NAME]-vpc`

File: cmd/infra/aws/create.go [hunk 17]
```diff
@@ -1,225 +1,394 @@
...
+		case cloudformation.StackStatusCreateFailed:
+			return false, fmt.Errorf("stack creation failed")
+		case cloudformation.StackStatusRollbackInProgress,
+			cloudformation.StackStatusRollbackComplete,
+			cloudformation.StackStatusRollbackFailed:
+			return false, fmt.Errorf("stack creation failed and was rolled back")
+		default:
+			return false, fmt.Errorf("unexpected stack creation status: %s", *stack.StackStatus)
+		}
+	}, ctx.Done())
```
Comment (enxebre) [inline cmd/infra/aws/create.go:262]: this is great now but still the error returned here will be a meaningless timeout message returned by `PollUntil`, it'd be useful to capture the one returned by your inlined function and pass it though here, something like:
```
var actualErr
err = wait.PollUntil(5*time.Second, func() (bool, error) {
  actualErr = 
} ctx.Done())
	if err != nil {
		return nil, fmt.Errorf("error getting stack: %v, %w", err, actualErr)
	}
```

wdyt?

File: cmd/infra/aws/cloudformation/cluster.yaml (hunk 4 above)
Comment (csrwng) [reply cmd/infra/aws/cloudformation/cluster.yaml:36]: Can we add all of the public ones? It's a finite list.

File: cmd/infra/aws/create.go (hunk 17 above)
Comment (ironcladlou) [reply cmd/infra/aws/create.go:262]: If `PollUtil` returns because of an error I returned inline (not a timeout), the useful error is propagated out already. If `PollUntil` returns because of the context closing (eg timeout), the timeout error generated by `PollUntil` is accurate, because the function didn't return due to the inline code returning any error — there is no actual error to wrap. Make sense? Maybe I'm misunderstanding.

File: cmd/infra/aws/destroy.go (hunk 10 above)
Comment (ironcladlou) [reply cmd/infra/aws/destroy.go:145]: I do agree NLB and some volume stuff is inevitable to cover the stuff we _know_ OCP will do by default, but apparently we're not yet exercising any of those features in our e2e clusters.

File: cmd/infra/aws/create.go (hunk 17 above)
Comment (enxebre) [reply cmd/infra/aws/create.go:262]: right, sorry I was forgetting and returning error finish the PollUntil right away.

File: cmd/infra/aws/cloudformation/cluster.yaml (hunk 4 above)
Comment (ironcladlou) [reply cmd/infra/aws/cloudformation/cluster.yaml:39]: I went ahead and added the US ones, there might be a way to eliminate this mapping entirely so I added a TODO

File: cmd/infra/aws/cloudformation/cluster.yaml (hunk 5 above)
Comment (ironcladlou) [reply cmd/infra/aws/cloudformation/cluster.yaml:59]: I went through and added what seem like reasonable Name tags to every tagged resource, PTAL

Comment (csrwng) [reply cmd/infra/aws/cloudformation/cluster.yaml:59]: lgtm
